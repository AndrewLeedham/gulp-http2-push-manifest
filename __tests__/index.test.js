const gulp = require('gulp');
const path = require('path');
const fs = require('fs');
const through = require('through2');

const consoleLogSpy = jest
  .spyOn(global.console, 'log')
  .mockImplementation(() => {});
const writeFileSyncSpy = jest
  .spyOn(fs, 'writeFileSync')
  .mockImplementation(() => {});
const manifest = require('..');

// Adapted from https://github.com/gulp-community/gulp-concat/blob/master/test/main.js#L12
function fixtures(glob) {
  return path.join(__dirname, '__fixtures__', glob);
}

// Inspired by https://github.com/floatdrop/stream-assert/blob/master/index.js#L49.
function end() {
  return through.obj((chunk, enc, callback) => callback());
}

async function run(file, options = undefined) {
  await new Promise((resolve) => {
    gulp
      .src(fixtures(file))
      .pipe(manifest(options))
      .pipe(end())
      .on('finish', resolve);
  });
}

beforeEach(async () => {
  jest.clearAllMocks();
});

for (const verbose of [true, false]) {
  const v = verbose ? ' - verbose' : '';
  const files = [
    { name: 'zero', number: 0 },
    { name: 'one', number: 1 },
    { name: 'two', number: 2 }
  ];
  for (const { name, number } of files) {
    test(`${number} manifest resources${v}`, async () => {
      await run(`${name}.html`, { verbose });

      expect(writeFileSyncSpy).toHaveBeenCalledWith(
        'push_manifest.json',
        expect.any(String)
      );
      expect(writeFileSyncSpy.mock.calls[0][1]).toMatchSnapshot();
      if (verbose) {
        expect(consoleLogSpy).toHaveBeenCalledTimes(number + 1);
        expect(consoleLogSpy).toHaveBeenCalledWith(
          'gulp-http2-push-manifest: ',
          `Found ${number} resource URLs in ${name}.html:`
        );
      } else {
        expect(consoleLogSpy).not.toHaveBeenCalled();
      }
    });
  }
}

test('custom manifest name', async () => {
  await run('two.html', { manifestName: 'custom.json' });
  expect(writeFileSyncSpy).toHaveBeenCalledWith(
    'custom.json',
    expect.any(String)
  );
  expect(writeFileSyncSpy.mock.calls[0][1]).toMatchSnapshot();
});
