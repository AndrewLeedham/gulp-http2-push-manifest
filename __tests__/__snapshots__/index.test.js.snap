// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0 manifest resources - verbose 1`] = `"{}"`;

exports[`0 manifest resources 1`] = `"{}"`;

exports[`1 manifest resources - verbose 1`] = `
"{
  \\"/styles.css\\": {
    \\"weight\\": 1,
    \\"type\\": \\"style\\"
  }
}"
`;

exports[`1 manifest resources 1`] = `
"{
  \\"/styles.css\\": {
    \\"weight\\": 1,
    \\"type\\": \\"style\\"
  }
}"
`;

exports[`2 manifest resources - verbose 1`] = `
"{
  \\"/styles.css\\": {
    \\"weight\\": 1,
    \\"type\\": \\"style\\"
  },
  \\"/script.js\\": {
    \\"weight\\": 1,
    \\"type\\": \\"script\\"
  }
}"
`;

exports[`2 manifest resources 1`] = `
"{
  \\"/styles.css\\": {
    \\"weight\\": 1,
    \\"type\\": \\"style\\"
  },
  \\"/script.js\\": {
    \\"weight\\": 1,
    \\"type\\": \\"script\\"
  }
}"
`;

exports[`custom manifest name 1`] = `
"{
  \\"/styles.css\\": {
    \\"weight\\": 1,
    \\"type\\": \\"style\\"
  },
  \\"/script.js\\": {
    \\"weight\\": 1,
    \\"type\\": \\"script\\"
  }
}"
`;
